### PR TITLE
build: add coverage-on-changed-files check to pre-push gate

### PR DESCRIPTION
**Pull request by sydlexius**
## Summary

- Generates a cover profile during the existing `go test -race` run (zero extra test execution cost -- same run, added flags)
- After tests pass, checks whether any function in a changed source file shows 0% coverage
- Prints a `WARNING` with the uncovered function list and a Codecov note
- Non-blocking (exits 0) -- functions intentionally tested via integration don't become hard blockers

## Motivation

PR #1035 had a `codecov/patch` failure caught only after CI. This check would have surfaced the gap at pre-push time, prompting the test addition before the PR was opened.

## Test plan
- [ ] Run `bash scripts/pre-push-gate.sh` on a branch with an uncovered changed function -- verify WARNING appears
- [ ] Run on a branch with no Go source changes -- verify "Skipped" message
- [ ] Run on a branch with covered changed functions -- verify "OK"

Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Enhanced pre-push validation to generate coverage data and warn when any modified Go function has 0% coverage, skipping when no relevant changes or coverage cannot be parsed.
* **Chores**
  * Consolidated temporary artifact cleanup into a single shared teardown to ensure reliable removal of coverage and API temp files.

### CODE REVIEW

File: scripts/pre-push-gate.sh [hunk 1]
```diff
@@ -6,8 +6,11 @@ set -euo pipefail
 SCRIPT_DIR="$(cd "$(dirname "$0")" && pwd)"
 BASE=$(git merge-base main HEAD 2>/dev/null || echo "HEAD~1")
 
+COVER_OUT=$(mktemp /tmp/stillwater-cover.XXXXXX.out)
+trap 'rm -f "${COVER_OUT:-}"' EXIT
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply scripts/pre-push-gate.sh]: Fixed in e0eaadd. Consolidated the two EXIT traps into a single cleanup() function (with tmp_openapi pre-declared as empty) so COVER_OUT cleanup is no longer dropped when oasdiff is available. Also removed the redundant inline rm calls inside the oasdiff block since cleanup() now owns all temp file removal.

File: scripts/pre-push-gate.sh [hunk 2]
```diff
@@ -54,5 +57,29 @@ else
...
 
+echo ""
+echo "=== Coverage on changed files ==="
+changed_go=$(git diff "$BASE"..HEAD --name-only -- '*.go' | grep -v '_test.go' || true)
+if [ -z "$changed_go" ]; then
+  echo "Skipped (no Go source changes)."
+else
+  uncovered=""
+  while IFS= read -r f; do
+    hits=$(go tool cover -func="$COVER_OUT" 2>/dev/null | grep "/${f}:" | awk '$NF == "0.0%"' || true)
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply scripts/pre-push-gate.sh]: Fixed in e0eaadd. Hoisted go tool cover -func into cover_funcs before the loop; each iteration now filters the cached string with printf | grep instead of re-invoking the cover tool.

**Comment by sydlexius** [comment]
@coderabbitai resolve

File: scripts/pre-push-gate.sh [hunk 3]
```diff
@@ -36,23 +43,45 @@ echo ""
...
+  done <<< "$changed_go"
+
+  if [ -n "$uncovered" ]; then
+    echo "WARNING: Functions with 0% coverage in changed files:"
+    echo "$uncovered"
+    echo "These may be intentionally tested via integration or missing unit tests."
+    echo "Codecov will flag patch coverage -- consider adding tests before opening a PR."
+  else
+    echo "OK"
+  fi
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply scripts/pre-push-gate.sh]: Fixed in aa4c746. Replaced the || true suppression with an explicit failure check that emits a warning and skips the check when go tool cover fails (preventing a false OK). Also switched to grep -F so filenames are matched as fixed strings rather than regex patterns.

**Comment by sydlexius** [comment]
@coderabbitai resolve

**Comment by sydlexius** [comment]
@coderabbitai status
